### PR TITLE
RDoc-798 documenting bulk insert

### PR DESCRIPTION
**Pull request by talweiss1982**

### CODE REVIEW

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+| Return Value | |
+| ------------- | ----- |
+| [BulkInsertOperation](../../glossary/bulk-insert-operation) | Instance of BulkInsertOperation used for interaction. |
+
+## Limitations
+
+There are several limitations to the API:
+
+* Entity **Id** must be provided by the client. The client by default will use the HiLo generator in order to generate the **Id**.
+* Transactions are per batch, not per operation.
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Explains what this means, that there is not a single transaction for the whole thing.

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+* Documents inserted using bulk-insert will not raise notifications. More about `Changes API` can be found [here](../changes/what-is-changes-api).
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: They will

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+* Bulk insert is not thread safe, it should not be accessed concurrently.
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Mention that we can have concurrent bulk inserts, but not access the same one concurrently.

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+| Return Value | |
+| ------------- | ----- |
+| [BulkInsertOperation](../../glossary/bulk-insert-operation) | Instance of BulkInsertOperation used for interaction. |
+
+## Limitations
+
+There are several limitations to the API:
+
+* Entity **Id** must be provided by the client. The client by default will use the HiLo generator in order to generate the **Id**.
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Can drop this, IIRC, we do support it now, just not idenities

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/.docs.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿[
+    {
+        "Path": "how-to-work-with-bulk-insert-operation.markdown",
+        "Name": "How to work with bulk insert operation?",
+        "Mappings": [
+            {
+                "Version": 3.5,
+                "Key": "client-api/bulk-insert"
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/.docs.json]: FYI. If the path is the same in 3.5 then you do not need that.

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Raven.Documentation.Samples.csproj [hunk 6]
```diff
@@ -5,10 +5,16 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="RavenDB.Client" Version="4.0.0-nightly-20171204-0501" />
+    <PackageReference Include="RavenDB.Client" Version="3.0.30171" />
```
Comment (ppekrol) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Raven.Documentation.Samples.csproj]: ?

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Raven.Documentation.Samples.csproj [hunk 7]
```diff
@@ -5,10 +5,16 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="RavenDB.Client" Version="4.0.0-nightly-20171204-0501" />
+    <PackageReference Include="RavenDB.Client" Version="3.0.30171" />
     <PackageReference Include="xunit" Version="2.3.1" />
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (ppekrol) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/Raven.Documentation.Samples.csproj]: ?

File: Documentation/4.0/Raven.Documentation.Pages/client-api/changes/.docs.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿[
+    {
+        "Path": "what-is-changes-api.markdown",
+        "Name": "What is Changes API?",
+        "Mappings": [
+            {
+                "Version": 3.5,
+                "Key": "client-api/changes"
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/changes/.docs.json]: FYI. If the path is the same in 3.5 then you do not need that.

File: Documentation/4.0/Raven.Documentation.Pages/glossary/bulk-insert-operation.markdown [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿# Glossary : BulkInsertOperation
+
+### Methods
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/glossary/bulk-insert-operation.markdown]: Let's maybe integrate this with the article

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+## Limitations
+
+There are several limitations to the API:
+
+* The bulk insert operation is broken into batches, each batch is treated in its own transaction so the whole operation isn't treated under a single transaction.
+* Bulk insert is not thread safe, a single bulk insert should not be accessed concurrently. The use of multiple bulk inserts, on the same client, concurrently is supported.
+
+## Example
+
+{CODE bulk_inserts_4@ClientApi\BulkInsert\BulkInserts.cs /}
```
Comment (ppekrol) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Let's add Code-Tabs here with Sync and Async example

File: Documentation/4.0/Raven.Documentation.Pages/Raven.Documentation.Pages.csproj [hunk 11]
```diff
@@ -42,23 +42,19 @@
...
+    <None Include="client-api\bulk-insert\.docs.json" />
+    <None Include="client-api\bulk-insert\how-to-work-with-bulk-insert-operation.markdown" />
+    <None Include="client-api\changes\.docs.json" />
+    <None Include="client-api\changes\what-is-changes-api.dotnet.markdown" />
     <None Include="client-api\commands\documents\get.dotnet.markdown" />
     <None Include="client-api\commands\documents\how-to\get-document-metadata-only.dotnet.markdown" />
     <None Include="client-api\creating-document-store.markdown" />
     <None Include="client-api\data-subscriptions\.docs.json" />
     <None Include="client-api\data-subscriptions\how-to-create-data-subscription.dotnet.markdown" />
-    <None Include="client-api\data-subscriptions\how-to-consume-data-subscription.markdown" />
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/Raven.Documentation.Pages.csproj]: Was the removal of all of these files intentional?

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+One of the features that is particularly useful when inserting large amount of data is `bulk inserting`. This is an optimized time-saving approach with few drawbacks that will be described later.
+
+## Syntax
+
+{CODE bulk_inserts_1@ClientApi\BulkInsert\BulkInserts.cs /}
+
+| Parameters | | |
+| ------------- | ------------- | ----- |
+| **database** | string | Name of database for which bulk operation should be performed. If `null` then `DefaultDatabase` from DocumentStore will be used. |
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: There is no `DefaultDatabase` property, just `Database`

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+| **void StoreAsync(object entity, IMetadataDictionary metadata = null)** | store the entity in an async manner, identifier will be generated automatically on client-side. Optional, metadata can be provided for the stored entity. |
+| **void StoreAsync(object entity, string id, IMetadataDictionary metadata = null)** | store the entity in an async manner, with `id` parameter to explicitly declare the entity identifier. Optional, metadata can be provided for the stored entity.|
+| **void Dispose()** | Dispose an object |
+
+## Limitations
+
+There are several limitations to the API:
+
+* The bulk insert operation is broken into batches, each batch is treated in its own transaction so the whole operation isn't treated under a single transaction.
+* Bulk insert is not thread safe, a single bulk insert should not be accessed concurrently. The use of multiple bulk inserts, on the same client, concurrently is supported.
```
Comment (ayende) [inline Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Need to mention that it is async safe, just to be 100% complete here

File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/BulkInsert/BulkInserts.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+                using (BulkInsertOperation bulkInsert = store.BulkInsert())
+                {
+                    for (int i = 0; i < 1000 * 1000; i++)
+                    {
+                        await bulkInsert.StoreAsync(new Employee
+                        {
+                            FirstName = "FirstName #" + i,
+                            LastName = "LastName #" + i
+                        });
+                    }
```
Comment (ayende) [inline Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/BulkInsert/BulkInserts.cs]: add `DisposeAsync` here

File: Documentation/4.0/Raven.Documentation.Pages/Raven.Documentation.Pages.csproj (hunk 11 above)
Comment (talweiss1982) [reply Documentation/4.0/Raven.Documentation.Pages/Raven.Documentation.Pages.csproj]: no probably merge issue

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown (hunk 13 above)
Comment (talweiss1982) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: what do you mean by async safe?

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown (hunk 12 above)
Comment (talweiss1982) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: k

File: Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown (hunk 13 above)
Comment (ayende) [reply Documentation/4.0/Raven.Documentation.Pages/client-api/bulk-insert/how-to-work-with-bulk-insert-operation.markdown]: Someone not familiar with how async work may assume that you can't use bulk insert in an async context.

**Pull request merged**